dts: edtlib: Support giving missing properties a default value

ulfalizer opened the pull request:
For missing optional properties, it can be handy to generate a default
value instead of no value, to cut down on #ifdefs.

Allow a default value to be specified in the binding, via a new
'default: <default value>' setting for properties in bindings.
Defaults are supported for both scalar and array types. YAML arrays are
used to specify the value for array types.

'default:' also appears in json-schema, with the same meaning.

Include misc. sanity checks, like the 'default' value matching 'type'.

The documentation changes in binding-template.yaml explain the syntax.

Suggested by Peter A. Bigot in
https://github.com/zephyrproject-rtos/zephyr/issues/17829.

Fixes: #17829

ulfalizer committed:
dts: edtlib: Support giving missing properties a default value

For missing optional properties, it can be handy to generate a default
value instead of no value, to cut down on #ifdefs.

Allow a default value to be specified in the binding, via a new
'default: <default value>' setting for properties in bindings.
Defaults are supported for both scalar and array types. YAML arrays are
used to specify the value for array types.

'default:' also appears in json-schema, with the same meaning.

Include misc. sanity checks, like the 'default' value matching 'type'.

The documentation changes in binding-template.yaml explain the syntax.

Suggested by Peter A. Bigot in
https://github.com/zephyrproject-rtos/zephyr/issues/17829.

Fixes: #17829
Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer commented:
This PR is currently bundled on top of https://github.com/zephyrproject-rtos/zephyr/pull/18766, so that I could use the new `compatible` syntax in the tests. It's a simple dependency though.

dbkinder reviewed (approved):
doc change LGTM, but it's the same change found in https://github.com/zephyrproject-rtos/zephyr/pull/18766 (so no conflict).

galak commented on dts/binding-template.yaml:
```diff
@@ -125,6 +126,15 @@ sub-node:
+# given for a property <prop> and <prop> is missing, then the output will be as
+# if '<prop> = <default>' had appeared (except YAML data types are used for the
+# default value). Note that it only makes sense to combine 'default:' with
+# 'required: false'.
```
Do we error check that `default:` is only set for `required: false`?

ulfalizer replied:
I'll add a check.

ulfalizer replied:
Hmm... wonder if this might be too strict though. If a binding changes a property from an included binding that happens to have has a `default:` from `required: false` to `required: true`, then it might be bad to flag that as an error.

I'll leave it out unless you think it should be flagged.

galak replied:
if we inherit in such a case, then `required` goes from `false` -> `true`, then the `default` should not propagate.

I'd rather us add the check and see if such a case comes up.

ulfalizer replied:
Alright, I'll add the check.

I'll skip stripping the `default` for now. Could see later if it's actually needed.

ulfalizer replied:
Added check and updated documentation.

pabigot replied:
At one point I thought I suggested that `default` be used only if `required`.  It doesn't make sense to have `optional` with a default because the motivation of `optional` is to indicate that something may not be provided at all, in which case the code probably does something special.

galak replied:
@pabigot not sure I follow.

galak commented on dts/binding-template.yaml:
```diff
@@ -158,6 +168,31 @@ properties:
 
+    int-with-default:
+        type: int
+        category: optional
```
Examples need updating for `required` instead of `category`.

ulfalizer replied:
Will fix.

ulfalizer replied:
Fixed

galak reviewed (commented):
We should comment that `default:` is only valid for `string | int | array | uint8-array | string-array` type(s).  (and ideally error check that).

ulfalizer commented:
> We should comment that `default:` is only valid for `string | int | array | uint8-array | string-array` type(s). (and ideally error check that).

Already got a check for it:

```python
if prop_type in {"boolean", "compound", "phandle", "phandles",
                 "phandle-array"}:
    _err("'default:' can't be combined with 'type: {}' for '{}' in "
         "'properties:' in {}".format(prop_type, prop_name, binding_path))
```

Nice to mention in the documentation as well though, so I'll add it.

ulfalizer commented:
Marking DNM as https://github.com/zephyrproject-rtos/zephyr/pull/18766 has been updated and needs to go in first, just to unconfuse myself.

Will rebase this one later.

ulfalizer commented:
@galak 
Rebased the PR and addressed all review comments.

See the comment above re. `required: true` and `default:`. Realized it might be too strict.

erwango commented on dts/binding-template.yaml:
```diff
@@ -154,6 +166,31 @@ properties:
+        type: uint8-array
+        required: false
+        default: [0x12, 0x34] # Same as 'uint8-array-with-default = [12 34]'
+
```
no `boolean-with-default` ?
I see you took care of it in the lib, so this comment would just be about documenting it.

ulfalizer replied:
If it's handled, it's unintentional. ;)

Putting `default:` on a `boolean` gives an error:

```python
if prop_type in {"boolean", "compound", "phandle", "phandles",
                 "phandle-array"}:
    _err("'default:' can't be combined with 'type: {}' for '{}' in "
         "'properties:' in {}".format(prop_type, prop_name, binding_path))
```

Not sure what it would mean to put a default on a boolean. `default: true` seems weird, because then there'd be no way to check if the property is there (would get `true` both if it's there and if it's missing), and `default: false` would be a no-op, since that's already the value if  the property is missing.

Note that the `default:` value has to match the type.

pabigot replied:
Yeah, this fits in with my comment above.  `boolean` by its very model is always provided, as either 1 or 0.  Functionally it's as if `boolean` had an implicit `required` with implicit default `false`.  Only when a property is `optional` do you fail to get a generated symbol for it.

ulfalizer replied:
@erwango 
Ping

Anything about the current behavior you're unhappy with?

erwango replied:
@ulfalizer, my point was to care about `default: false`, but indeed that's already taken care of, so fine with me! Txs

erwango reviewed (changes requested):
Follow up on the boolean topic.
Even if boolean type is not impacted by this change, because it already has its own logic, a word about it (mentioning 'default' is not needed) would be good for people not aware about internal script logic.

ulfalizer committed:
dts: binding-template.yaml: Document how 'type: boolean' works

It's a bit subtle in that it's the only type where a property can
generate output even if it doesn't exist.

Signed-off-by: Ulf Magnusson <Ulf.Magnusson@nordicsemi.no>

ulfalizer commented:
> Follow up on the boolean topic.
> Even if boolean type is not impacted by this change, because it already has its own logic, a word about it (mentioning 'default' is not needed) would be good for people not aware about internal script logic.

Yeah, nice to have it documented. Added some documentation in a separate commit now (see the branch tip).
